artemis-server/src/main/resources/schema/artemis-configuration.xsd

```diff
@@ -16,18 +16,23 @@
   limitations under the License.
```

Comment:
Not sure whats going on with soo many changes here, the JIRA mentions diverts, this is already supported to be able to include in a deployed configuration setup (follow docs) i know many installs in the real world already doing this today.

Reply:
The intent of the changes is to remove all the embedded {{element}} and {{complexType}} definitions from the {{core}} definition.

By moving these schema definitions, XML tooling becomes valid for both included XML elements and the overall {{broker.xml}} file. As with the {{diverts}} example in the JIRA (not mine--just trying to address it), an IDE has no idea what a {{diverts}} is outside of a {{core}} element, meaning syntax help is not available. Any other valid imported element file, e.g. included files containing top-level {{addresses}}, {{connectors}}, {{acceptors}}, etc. elements, would also have the same syntax tooling support.

As you mentioned, XML includes work just fine currently, but the schema definition as it currently stands does not support included element files as well as it can and should.

Reply:
Im struggling here to see the issue, the elements are importable. The exact issue or want in the jira ticket having Diverts as a separate xml file is possible today.....

Reply:
Diverts is given as an example in the JIRA, not as the totality of required changes.

Reply:
But point is, modular is already supported and many key parts that most common users want out are... if theres a section that is sensible to be modular and isn't then we should focus on that, but this very large change (any large change is high risk, as you never know how it could impact others) so i would rather we focus on just what isn't possible to be modularised today, if its sensible to do so.

Reply:
And I understand your point. I work building large-scale production Artemis systems for clients every day--that's my job. I use modular files for common sections of broker.xml configurations extensively to build everything from cluster prototypes to multi-DC, world-wide messaging infrastructure.

My point is simple. The XSD does not support modularity as well as it should due to the "anonymous" element and complexType definitions buried in the configurationType definition. This is bad XSD form as the embedded defined types and elements cannot be used directly--they're only useful in the context of the outermost defining type, the configurationType in this case.

Moving these embedded elements and their associated complex types to top-level definitions enables XML tooling to recognize and support the schema definition in both modular files as they're being developed as well as complete broker.xml files. While the elements may not make sense outside a broker configuration, supporting modular files necessarily implies the elements therein are portable to those files. The same tooling support should be available in the modular files as is available when developing a monolithic broker.xml file.

I agree the scale of this change involves significant risk. This is why I included additional updates to the existing xinclude unit tests to address that risk. If you would like to suggest additional test scenarios, I'm happy to include those as well.

Reply:
As i said, if theres a specific, area that needs to change great, im happy to help focus to address the point area, but as it stands and as a project maintainer i don't feel that comfortable with this change the risk is very large and i don't feel is needed personally. If other maintainers (members) wish to jump in here, and feel very confident about the change not breaking anyone's setups then great.

Reply:
If you would like me to replicate the original test files with an alternate modular include unit test scenario, I will do so. All the original test files work unaltered with the updated schema definition, as evidenced by the broker-connection and security-settings files. The existing config changes were to avoid creating a separate unit test class, which I will now create and submit.

The attached screenshot shows one example of the IDE syntax support these schema definition changes enable when developing individual modular files, not just broker.xml files.

![XSDIncludeSyntaxSupport](https://user-images.githubusercontent.com/21245081/176733040-16ac65e0-b111-4937-9bff-9600289f4895.png)

I would welcome other input here as well.